### DISCUSSION TITLE
std::string + vs absl::StrCat time cost

### DISCUSSION THREAD

**Discussion by zqw-hooper** [abseil/abseil-cpp#1754] [Q&A]
![image](https://github.com/user-attachments/assets/4b15c981-4353-4ef2-8c9a-949291df8631)

**Comment by zqw-hooper** [comment]
why absl::StrCat has much more time cost than std::string ？ compile with -O3 opti

**Reply by derekmauro** [reply]
Please do not post screenshots of code. Copy and paste the code so that if someone wants to try it, they don't have to retype it.

Writing benchmarks is tricky with optimizing compilers. I didn't try this myself, but `result` and `result2` are unused variables, which means the compiler is allowed to optimize them away. If, for example, the standard library code for `s1 + s2` is entirely in a header, the compiler may be able to see there are no side-effects to `s1 + s2`, and thus does not actually have to execute the code at all. `absl::StrCat()`, on the other hand, is not header-only, and involves a call into another library. The compiler cannot see everything it does in your translation unit, and therefore cannot know that it has no side-effects, and thus must execute the code.

So in short, I don't believe your benchmark is actually running  `s1 + s2` at all.

Try writing your benchmark using the [Google Benchmark](https://github.com/google/benchmark) framework and calling `benchmark::DoNotOptimize()` on `result` and `result2`.

**Reply by zqw-hooper** [reply]
```cpp
#include <string>
#include <benchmark/benchmark.h>
#include <iostream>
#include "absl/strings/str_cat.h"

static void std_string_test(benchmark::State& state)
{
  std::string s1 = "Hello, ";
  std::string s2 = "World!";
  std::string result;
  for (auto _ : state)
  {
    result = s1 + s2;
    benchmark::DoNotOptimize(result);
  }
}

BENCHMARK(std_string_test);

// Define another benchmark
static void absl_string_test(benchmark::State& state)
{
  std::string s1 = "Hello, ";
  std::string s2 = "World!";
  std::string result2;
  for (auto _ : state)
  {
    result2 = absl::StrCat(s1, s2);
    benchmark::DoNotOptimize(result2);
  }
}

BENCHMARK(absl_string_test);

BENCHMARK_MAIN();
```

and test reuslt with benchmark, **why abseil-cpp has bad performance**:  
![image](https://github.com/user-attachments/assets/2fb27f01-662a-42d2-9d03-d92aea342c8b)

**Reply by zqw-hooper** [reply]
```cpp
#include <string>
#include <benchmark/benchmark.h>
#include <iostream>
#include "absl/strings/str_cat.h"
#include "absl/strings/str_format.h"

static void std_string_test(benchmark::State& state)
{
  std::string s1 = "String ";
  std::string s2 = "merge test: ";
  int test_int = 0;
  double test_double = 0.0;

  std::string result;
  for (auto _ : state)
  {
    test_int++;
    test_double += 0.1;
    result =
      s1 + s2 + "and test_int = " + std::to_string(test_int) + " and test_double = " + std::to_string(test_double);
    benchmark::DoNotOptimize(result);
  }
  std::cout << result << std::endl;
}

BENCHMARK(std_string_test);

// Define another benchmark
static void absl_string_test(benchmark::State& state)
{
  std::string s1 = "String ";
  std::string s2 = "merge test: ";
  int test_int = 0;
  double test_double = 0.0;

  std::string result2;
  for (auto _ : state)
  {
    test_int++;
    test_double += 0.1;
    result2 =
      absl::StrCat(s1, s2, "and test_int = ", test_int, " and test_double = ", absl::StrFormat("%.6f", test_double));
    benchmark::DoNotOptimize(result2);
  }
  std::cout << result2 << std::endl;
}

BENCHMARK(absl_string_test);

BENCHMARK_MAIN();
```
and tested hybrid with string、int and double，test result also abseil has bad performance:
![image](https://github.com/user-attachments/assets/804aaefa-bbe6-42f5-8456-61458e1a46e9)

**Reply by derekmauro** [reply]
This is what I get. I suspect your build is configured poorly.

```
$ bazel-bin/absl/strings/str_cat_benchmark
2024-09-15T20:39:55-04:00
Running bazel-bin/absl/strings/str_cat_benchmark
Run on (8 X 4800 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
Load Average: 3.43, 1.66, 0.85
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
String merge test: and test_int = 1 and test_double = 0.100000
String merge test: and test_int = 10 and test_double = 1.000000
String merge test: and test_int = 100 and test_double = 10.000000
String merge test: and test_int = 1000 and test_double = 100.000000
String merge test: and test_int = 10000 and test_double = 1000.000000
String merge test: and test_int = 100000 and test_double = 10000.000000
String merge test: and test_int = 826981 and test_double = 82698.100000
-----------------------------------------------------------
Benchmark                 Time             CPU   Iterations
-----------------------------------------------------------
std_string_test         788 ns          788 ns       826981
String merge test: and test_int = 1 and test_double = 0.100000
String merge test: and test_int = 10 and test_double = 1.000000
String merge test: and test_int = 100 and test_double = 10.000000
String merge test: and test_int = 1000 and test_double = 100.000000
String merge test: and test_int = 10000 and test_double = 1000.000000
String merge test: and test_int = 100000 and test_double = 10000.000000
String merge test: and test_int = 1000000 and test_double = 100000.000001
String merge test: and test_int = 2647446 and test_double = 264744.600010
absl_string_test        262 ns          261 ns      2647446
```

**Reply by zqw-hooper** [reply]
i found the problem because my local abseil-cpp compile without optimizations, and i replan it with "-O3" level, and result is below: 
![image](https://github.com/user-attachments/assets/4fda41c1-335a-4ee4-a411-c7068a1cd262)
